aptos-move/aptos-vm/src/move_vm_ext/session.rs

```diff
@@ -92,18 +97,111 @@ impl SessionId {
 
 pub struct SessionExt<'r, 'l, S> {
     inner: Session<'r, 'l, S>,
+    remote: MoveResolverWithVMMetadata<'r, 'l, S>,
 }
 
 impl<'r, 'l, S> SessionExt<'r, 'l, S>
 where
     S: MoveResolverExt,
 {
-    pub fn new(inner: Session<'r, 'l, S>) -> Self {
-        Self { inner }
+    pub fn new(inner: Session<'r, 'l, S>, move_vm: &'l MoveVM, remote: &'r S) -> Self {
+        Self {
+            inner,
+            remote: MoveResolverWithVMMetadata::new(remote, move_vm),
+        }
     }
 
     pub fn finish(self) -> VMResult<SessionOutput> {
         let (change_set, events, mut extensions) = self.inner.finish_with_extensions()?;
+
+        let common_error = PartialVMError::new(StatusCode::UNREACHABLE).finish(Location::Undefined);
+        let mut change_set_grouped = MoveChangeSet::new();
+        for (addr, account_changeset) in change_set.into_inner() {
+            let mut resource_groups: BTreeMap<StructTag, AccountChangeSet> = BTreeMap::new();
+            let mut account_changeset_grouped = AccountChangeSet::new();
+            let (modules, resources) = account_changeset.into_inner();
+
+            // * Separate the resource groups from the non-resource groups
+            // * non-resource groups are kept as is
+            // * resource groups are merged into the correct format as deltas to the source data
+            for (struct_tag, blob_op) in resources {
+                let resource_group = self.remote.get_resource_group(&struct_tag).map_err(|_| common_error.clone())?;
+                if let Some(resource_group) = resource_group {
+                    resource_groups
+                        .entry(resource_group)
+                        .or_insert_with(AccountChangeSet::new)
+                        .add_resource_op(struct_tag, blob_op)
+                        .map_err(|_| common_error.clone())?;
+                    continue;
```

Comment:
nit: I think having an else statement here would make it clearer that we're either inserting the op into the right resource group's changeset in the BTreeMap or in the global resource changeset.

Reply:
personally I prefer continues over elses :p